src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -176,6 +176,133 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
+            {
+                addCompatibleExtension(binder, extension, receiverType, compatibleExtensions, ref useSiteInfo);
+            }
+
+            return;
+
+            static void addCompatibleExtension(Binder binder, NamedTypeSymbol extension, TypeSymbol receiverType, ArrayBuilder<NamedTypeSymbol> compatibleExtensions, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                if (extension.ExtensionParameter is not { } extensionParameter)
+                {
+                    return;
+                }
+
+                var compilation = binder.Compilation;
+                var constructedExtension = inferExtensionTypeArguments(extension, receiverType, compilation, ref useSiteInfo);
+                if (constructedExtension is null)
+                {
+                    return;
+                }
+
+                Debug.Assert(constructedExtension.ExtensionParameter is not null);
+                var conversion = compilation.Conversions.ConvertExtensionMethodThisArg(parameterType: constructedExtension.ExtensionParameter.Type, receiverType, ref useSiteInfo, isMethodGroupConversion: false);
+                if (!conversion.Exists)
+                {
+                    return;
+                }
+
+                compatibleExtensions.Add(constructedExtension);
+            }
+
+            static NamedTypeSymbol? inferExtensionTypeArguments(NamedTypeSymbol extension, TypeSymbol receiverType, CSharpCompilation compilation, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                if (extension.Arity == 0)
+                {
+                    return extension;
+                }
+
+                var containingAssembly = extension.ContainingAssembly;
+                var conversions = containingAssembly.CorLibrary.TypeConversions;
+
+                // Note: we create a value for purpose of inferring type arguments even when the receiver type is static
+                var syntax = (CSharpSyntaxNode)CSharpSyntaxTree.Dummy.GetRoot();
+                var receiverValue = new BoundLiteral(syntax, ConstantValue.Bad, receiverType) { WasCompilerGenerated = true };
+
+                var typeArguments = MethodTypeInferrer.InferTypeArgumentsFromReceiverType(extension, receiverValue, compilation, conversions, ref useSiteInfo);
+                if (typeArguments.IsDefault || typeArguments.Any(t => !t.HasType))
+                {
+                    return null;
+                }
+
+                bool success = checkConstraints(extension, extension.TypeParameters, typeArguments, compilation, conversions, ref useSiteInfo);
+                if (!success)
+                {
+                    return null;
+                }
+
+                return extension.Construct(typeArguments);
+            }
+
+            static bool checkConstraints(Symbol symbol, ImmutableArray<TypeParameterSymbol> typeParams, ImmutableArray<TypeWithAnnotations> typeArgs,
```

Comment:
>ImmutableArray<TypeParameterSymbol> typeParams [](http://example.com/codeflow?start=56&length=46)

It looks like this parameter is redundant, we can get type parameters from extension type.